tests/unit/test_hnswlib.cpp

```diff
@@ -614,3 +614,186 @@ TEST_F(HNSWLibTest, hnsw_override) {
+
+    // get the 100 vectors whose ids are the maximal among those that hasn't been returned yet, in
+    // every iteration. run this flow for 3 times, each time for 5 iteration, and reset the
+    // iterator.
+    size_t n_res = 100;
+    size_t total_iteration = 5;
+    size_t re_runs = 3;
+
+    for (size_t take = 0; take < re_runs; take++) {
+        size_t iteration_num = 0;
+        while (VecSimBatchIterator_HasNext(batchIterator)) {
+            std::vector<size_t> expected_ids(n_res);
+            for (size_t i = 0; i < n_res; i++) {
+                expected_ids[i] = (n - iteration_num * n_res - i - 1);
+            }
+            auto verify_res = [&](int id, float score, size_t index) {
+                ASSERT_TRUE(expected_ids[index] == id);
+            };
+            runBatchIteratorSearchTest(batchIterator, n_res, verify_res, BY_SCORE);
+            iteration_num++;
+            if (iteration_num == total_iteration) {
+                break;
+            }
+        }
+        VecSimBatchIterator_Reset(batchIterator);
+    }
+    VecSimBatchIterator_Free(batchIterator);
+    VecSimIndex_Free(index);
+}
+
+TEST_F(HNSWLibTest, hnsw_batch_iterator_advanced) {
+    size_t dim = 4;
+    size_t n = 1000;
+    size_t M = 8;
+    size_t ef = 20;
+
+    VecSimParams params = {.algo = VecSimAlgo_HNSWLIB,
+                           .hnswParams = {.type = VecSimType_FLOAT32,
+                                          .dim = dim,
+                                          .metric = VecSimMetric_L2,
+                                          .initialCapacity = n,
+                                          .M = M,
+                                          .efConstruction = 20,
+                                          .efRuntime = ef}};
+    VecSimIndex *index = VecSimIndex_New(&params);
+
+    float query[] = {(float)0, (float)0, (float)0, (float)0};
+    VecSimBatchIterator *batchIterator = VecSimBatchIterator_New(index, query);
+
+    // try to get results even though there are no vectors in the index.
+    VecSimQueryResult_List res = VecSimBatchIterator_Next(batchIterator, 10, BY_SCORE);
+    ASSERT_EQ(VecSimQueryResult_Len(res), 0);
+    VecSimQueryResult_Free(res);
+    ASSERT_FALSE(VecSimBatchIterator_HasNext(batchIterator));
+
+    // insert one vector and query
+    VecSimBatchIterator_Reset(batchIterator);
+    VecSimIndex_AddVector(index, query, 0);
+    res = VecSimBatchIterator_Next(batchIterator, 10, BY_SCORE);
+    ASSERT_EQ(VecSimQueryResult_Len(res), 1);
+    VecSimQueryResult_Free(res);
+    ASSERT_FALSE(VecSimBatchIterator_HasNext(batchIterator));
+
+    for (int i = 0; i < n; i++) {
+        float f[dim];
+        for (size_t j = 0; j < dim; j++) {
+            f[j] = (float)i;
+        }
+        VecSimIndex_AddVector(index, (const void *)f, i);
+    }
+    ASSERT_EQ(VecSimIndex_IndexSize(index), n);
+
+    // reset the iterator after it was depleted
+    VecSimBatchIterator_Reset(batchIterator);
```

Comment:
This wouldn't work, we must reset the iterator right before using it, otherwise the addition of vectors that comes between will increase the `tags` (and results won't be coherent).